iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/read/filescan/impl/UnclosedFileScanHandleImpl.java

```diff
@@ -0,0 +1,185 @@
+import org.apache.iotdb.db.storageengine.dataregion.read.filescan.model.AbstractChunkOffset;
+import org.apache.iotdb.db.storageengine.dataregion.read.filescan.model.AbstractDeviceChunkMetaData;
+import org.apache.iotdb.db.storageengine.dataregion.read.filescan.model.AlignedDeviceChunkMetaData;
+import org.apache.iotdb.db.storageengine.dataregion.read.filescan.model.ChunkOffset;
+import org.apache.iotdb.db.storageengine.dataregion.read.filescan.model.DeviceChunkMetaData;
+import org.apache.iotdb.db.storageengine.dataregion.tsfile.TsFileResource;
+import org.apache.iotdb.db.storageengine.dataregion.tsfile.timeindex.DeviceTimeIndex;
+import org.apache.iotdb.db.storageengine.dataregion.tsfile.timeindex.ITimeIndex;
+import org.apache.iotdb.db.storageengine.dataregion.utils.TsFileDeviceStartEndTimeIterator;
+import org.apache.iotdb.db.utils.ModificationUtils;
+
+import org.apache.tsfile.file.metadata.AlignedChunkMetadata;
+import org.apache.tsfile.file.metadata.IChunkMetadata;
+import org.apache.tsfile.file.metadata.IDeviceID;
+import org.apache.tsfile.file.metadata.statistics.Statistics;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.IntStream;
+
+public class UnclosedFileScanHandleImpl implements IFileScanHandle {
+
+  private final TsFileResource tsFileResource;
+  private final Map<IDeviceID, Map<String, List<IChunkMetadata>>> deviceToChunkMetadataMap;
+  private final Map<IDeviceID, Map<String, List<IChunkHandle>>> deviceToMemChunkHandleMap;
+
+  public UnclosedFileScanHandleImpl(
+      Map<IDeviceID, Map<String, List<IChunkMetadata>>> deviceToChunkMetadataMap,
+      Map<IDeviceID, Map<String, List<IChunkHandle>>> deviceToMemChunkHandleMap,
+      TsFileResource tsFileResource) {
+    this.deviceToChunkMetadataMap = deviceToChunkMetadataMap;
+    this.deviceToMemChunkHandleMap = deviceToMemChunkHandleMap;
+    this.tsFileResource = tsFileResource;
+  }
+
+  @Override
+  public TsFileDeviceStartEndTimeIterator getDeviceStartEndTimeIterator() throws IOException {
+    ITimeIndex timeIndex = tsFileResource.getTimeIndex();
+    return timeIndex instanceof DeviceTimeIndex
+        ? new TsFileDeviceStartEndTimeIterator((DeviceTimeIndex) timeIndex)
+        : new TsFileDeviceStartEndTimeIterator(tsFileResource.buildDeviceTimeIndex());
+  }
+
+  @Override
+  public boolean[] isDeviceTimeDeleted(IDeviceID deviceID, long[] timeArray) {
+    Map<String, List<IChunkMetadata>> chunkMetadataMap = deviceToChunkMetadataMap.get(deviceID);
+    boolean[] result = new boolean[timeArray.length];
+
+    chunkMetadataMap.values().stream()
+        .flatMap(List::stream)
+        .map(IChunkMetadata::getDeleteIntervalList)
+        .filter(deleteIntervalList -> !deleteIntervalList.isEmpty())
+        .forEach(
+            timeRangeList -> {
+              Integer deleteCursor = 0;
```

Comment:
this won't be changed in `ModificationUtils.isPointDeleted` and why you use `IntStream.range(0,timeArray.length).forEach` this way to iterate timeArray?